core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -133,6 +133,10 @@ private FileIO io(String location) {
     String impl = implFromLocation(location);
     FileIO io = ioInstances.get(impl);
     if (io != null) {
+      // re-apply the config in case it's null after Kryo serialization
+      if (io instanceof HadoopConfigurable && ((HadoopConfigurable) io).getConf() == null) {
```

Comment:
Should this operation be under the lock? It seems it is possible to modify concurrently.

Reply:
done

Reply:
Thanks @nastra. I have a question here: https://github.com/apache/iceberg/pull/3593#discussion_r1289498934. Do you think it should be a problem?

Reply:
not sure if this could be an issue, but we could try and switch `ioInstances` to a concurrent hash map instead of using locks. @ConeyLiu would you be interested in doing that?

Reply:
OK, I will take a try.